File: assembly/Style/block.ts
import * as u from "../Utils/index"; // Renamed to lowercase

const OPEN_CHARS: string[] = ['{', '[', '('];
const CLOSE_CHARS: string[] = ['}', ']', ')'];
const QUOTE_CHARS: string[] = ['`', "'", '"'];

export class ParseBlockResult { // Changed to export class
    adds: string[] = [];
    binds: string[] = [];
    preBinds: string[] = [];
    postBinds: string[] = [];
    variables: string[] = [];
    XatProps: Array<string[]> = []; // Changed to Array<string[]>
    atProps: Map<string, string> = new Map();
    Xproperties: Array<string[]> = []; // Changed to Array<string[]>
    properties: Map<string, string> = new Map();
    XatRules: Array<string[]> = [];    // Changed to Array<string[]>
    atRules: Map<string, string> = new Map();
    Xnested: Array<string[]> = [];      // Changed to Array<string[]>
    nested: Map<string, string> = new Map();
    Xclasses: Array<string[]> = [];     // Changed to Array<string[]>
    classes: Map<string, string> = new Map();
    Xflats: Array<string[]> = [];        // Changed to Array<string[]>
    flats: Map<string, string> = new Map();
    XallBlocks: Array<string[]> = [];   // Changed to Array<string[]>
    allBlocks: Map<string, string> = new Map();
}

export function parseBlock(content: string, blockArrays: boolean = false): ParseBlockResult {
    let result = new ParseBlockResult();
    content += ';';
    let keyStart: i32 = 0;
    let valStart: i32 = 0;
    let deviance: i32 = 0;
    let quote: string = '';
    let key: string = '';
    let isProp: boolean = true;
    let length: i32 = content.length;

    for (let index: i32 = 0; index < length; index++) {
        const ch: string = content.charAt(index);
        if (ch === '\\') {
            index++;
            continue;
        }
        if (QUOTE_CHARS.includes(ch)) {
            if (quote === '') {
                quote = ch;
            } else if (quote === ch) {
                quote = '';
            }
        }

        if (quote === '') {
            if (CLOSE_CHARS.includes(ch)) {
                deviance--;
            }

            if (deviance === 0) {
                switch (ch) {
                    case '{':
                        isProp = false;
                    case ':':
                        key = u.string.minify(content.substring(keyStart, index));
                        valStart = index + 1;
                        break;
                    case '}':
                    case ';':
                        const value: string = u.string.minify(content.substring(valStart, index));
                        if (isProp) {
                            if (key.length > 0) {
                                if (key.startsWith("--")) {
                                    result.variables.push(key);
                                }
                                result.properties.set(key, value);
                                if (blockArrays) result.Xproperties.push([key, value]);
                            } else if (value.length > 0 && value.charAt(0) === '@') { // added check for value
                                const spaceIndex: i32 = value.indexOf(" ");
                                const directive: string = value.substring(0, spaceIndex);
                                switch (directive) {
                                    case "@bind":
                                        result.binds.push(...u.string.zeroBreaks(value.substring(spaceIndex)));
                                        break;
                                    case "@pre-bind":
                                        result.preBinds.push(...u.string.zeroBreaks(value.substring(spaceIndex)));
                                        break;
                                    case "@post-bind":
                                        result.postBinds.push(...u.string.zeroBreaks(value.substring(spaceIndex)));
                                        break;
                                    case "@assemble":
                                        result.adds.push(...u.string.zeroBreaks(value.substring(spaceIndex)));
                                        break;
                                    default:
                                        result.atProps.set(value, "");
                                        if (blockArrays) result.XatProps.push([value, ""]);
                                }
                            } else {
                                const breaks: string[] = u.string.zeroBreaks(value);
                                const groupMap: Map<string, keyof ParseBlockResult> = new Map();
                                groupMap.set("<", "preBinds");
                                groupMap.set(">", "postBinds");
                                groupMap.set("+", "adds");
                                groupMap.set("*", "binds");
                                let group: keyof ParseBlockResult = groupMap.get(breaks[0]) || "adds";
                                if (groupMap.has(breaks[0])) {
                                    breaks.shift();
                                }
                                breaks.forEach(link => {
                                    const targetGroup: keyof ParseBlockResult = groupMap.get(link.charAt(0)) || group;
                                    result[targetGroup].push(groupMap.has(link.charAt(0)) ? link.substring(1) : link);
                                });
                            }
                        } else {
                            switch (key.charAt(0)) {
                                case "@":
                                    result.atRules.set(key, value);
                                    if (blockArrays) result.XatRules.push([key, value]);
                                    break;
                                case "&":
                                    result.nested.set(key, value);
                                    if (blockArrays) result.Xnested.push([key, value]);
                                    break;
                                case ".":
                                    result.classes.set(key, value);
                                    if (blockArrays) result.Xclasses.push([key, value]);
                                    break;
                                default:
                                    result.flats.set(key, value);
                                    if (blockArrays) result.Xflats.push([key, value]);
                            }
                            result.allBlocks.set(key, value);
                            if (blockArrays) result.XallBlocks.push([key, value]);
                        }
                        keyStart = index + 1;
                        valStart = index + 1;
                        key = '';
                        isProp = true;
                        break; // Added break here
                }
            }
            if (OPEN_CHARS.includes(ch)) {
                deviance++;
            }
        }
    }
    return result;
}

File: gosource/style/block.go
package style

import (
	_Cursor_ "main/Cursor"
	_cache_ "main/cache"
	_util_ "main/util"
	_slices_ "slices"
	"strings"
)

type block_Parse_retype struct{
	Assign      []string
	Attachment  []string
	Variables   map[string]string
	XatProps    [][2]string
	AtProps     map[string]string
	Xproperties [][2]string
	Properties  map[string]string
	XatRules    [][2]string
	AtRules     map[string]string
	Xnested     [][2]string
	Nested      map[string]string
	Xclasses    [][2]string
	Classes     map[string]string
	Xflats      [][2]string
	Flats       map[string]string
	XallBlocks  [][2]string
	AllBlocks   map[string]string
}

func block_Parse(content string, blockInArrays bool) block_Parse_retype {
	content += ";"

	result := block_Parse_retype{
		Assign:      []string{},
		Attachment:  []string{},
		Variables:   map[string]string{},
		XatProps:    [][2]string{},
		AtProps:     map[string]string{},
		Xproperties: [][2]string{},
		Properties:  map[string]string{},
		XatRules:    [][2]string{},
		AtRules:     map[string]string{},
		Xnested:     [][2]string{},
		Nested:      map[string]string{},
		Xclasses:    [][2]string{},
		Classes:     map[string]string{},
		Xflats:      [][2]string{},
		Flats:       map[string]string{},
		XallBlocks:  [][2]string{},
		AllBlocks:   map[string]string{},
	}

	key := ""
	quote := ' '
	keyStart := 0
	valStart := 0
	deviance := 0
	isProp := true
	cursor := _Cursor_.Construct(content + ";")

	for ch, streaming := cursor.Active.Char, cursor.Streaming; streaming; ch, streaming = cursor.Increment() {

		if ch == '\\' {
			cursor.Increment()
			continue
		}
		if _slices_.Contains(_cache_.Refer.WatchQuotes, ch) {
			switch quote {
			case ' ':
				quote = ch
			case ch:
				quote = ' '
			}
		}

		if quote == ' ' {
			if _slices_.Contains(_cache_.Refer.CloseBraces, ch) {
				deviance--
			}

			if deviance == 0 {
				switch ch {
				case '{':
					isProp = false
					key = _util_.String_Minify(content[keyStart:cursor.Active.Marker])
					valStart = cursor.Active.Marker + 1
				case ':':
					key = _util_.String_Minify(content[keyStart:cursor.Active.Marker])
					valStart = cursor.Active.Marker + 1
				case '}':
					fallthrough
				case ';':
					{
						value := _util_.String_Minify(content[valStart:cursor.Active.Marker])
						if isProp {
							if len(key) > 0 {
								if strings.HasPrefix(key, "--") {
									result.Variables[key] = value
								}
								result.Properties[key] = value
								if blockInArrays {
									result.Xproperties = append(result.Xproperties, [2]string{key, value})
								}
							} else if value[0] == '@' {
								spaceIndex := strings.Index(value, " ")
								if spaceIndex < 0 {
									spaceIndex = len(value)
								}
								directive := value[0:spaceIndex]

								switch directive {
								case _cache_.Root.CustomAtrules["attach"]:
									breaks := _util_.String_ZeroBreaks(value[spaceIndex:], []rune{' ', '\n', ','})
									result.Attachment = append(result.Attachment, breaks...)
								case _cache_.Root.CustomAtrules["assign"]:
									breaks := _util_.String_ZeroBreaks(value[spaceIndex:], []rune{' ', '\n', ','})
									result.Assign = append(result.Assign, breaks...)
								default:
									result.AtProps[value] = ""
									if blockInArrays {
										result.XatProps = append(result.XatProps, [2]string{value, ""})
									}
								}
							} else {
								breaks := _util_.String_ZeroBreaks(value, []rune{' ', '\n', ','})
								if breaks[0] == string(_cache_.Root.CustomOperations["attach"]) {
									result.Attachment = append(result.Attachment, breaks[1:]...)
								} else if (breaks[0]) == string(_cache_.Root.CustomOperations["assign"]) {
									result.Assign = append(result.Assign, breaks[1:]...)
								}
							}
						} else {
							switch rune(key[0]) {
							case '@':
								result.AtRules[key] = value
								if blockInArrays {
									result.XatRules = append(result.XatRules, [2]string{key, value})
								}
							case '&':
								result.Nested[key] = value
								if blockInArrays {
									result.Xnested = append(result.Xnested, [2]string{key, value})
								}
							case '.':
								result.Classes[key] = value
								if blockInArrays {
									result.Xclasses = append(result.Xclasses, [2]string{key, value})
								}
							default:
								result.Flats[key] = value
								if blockInArrays {
									result.Xflats = append(result.Xflats, [2]string{key, value})
								}
							}
							result.AllBlocks[key] = value
							if blockInArrays {
								result.XallBlocks = append(result.XallBlocks, [2]string{key, value})
							}
						}
						keyStart = cursor.Active.Marker + 1
						valStart = cursor.Active.Marker + 1
						key = ""
						isProp = true
					}
				}
			}

			if _slices_.Contains(_cache_.Refer.OpenBraces, ch) {
				deviance++
			}
		}
	}

	return result
}
